### PR TITLE
S3 Checkpoint loading from URI

### PR DESCRIPTION
**Pull request by eracah**
# What does this PR do?

Allow checkpoint loading for S3 without having to create a load_object_store

# What issue(s) does this change relate to?

fixes CO-1144

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [x] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/trainer/trainer.py [hunk 1]
```diff
@@ -304,6 +305,31 @@ def _is_tpu_installed() -> bool:
     import torch_xla.distributed.parallel_loader as pl
 
 
+def _maybe_create_object_store_from_uri(uri: str) -> Optional[ObjectStore]:
+    backend, bucket_name, _ = _parse_uri(uri)
+    if backend == '':
+        return None
+    if backend == 's3':
+        return S3ObjectStore(
+            bucket=bucket_name)  #bucket_uri=f'{backend}://{bucket_name}', backend_kwargs={'bucket': bucket_name})
```
Comment (dakinggg) [inline composer/trainer/trainer.py:314]: Remove comment

File: composer/trainer/trainer.py [hunk 2]
```diff
@@ -304,6 +305,31 @@ def _is_tpu_installed() -> bool:
...
+
+    elif backend == 'wandb':
+        raise NotImplementedError(f'There is no implementation for WandB load_object_store via URI. Please use '
+                                  'WandBLogger with log_artifacts set to True')
```
Comment (dakinggg) [inline composer/trainer/trainer.py:318]: For the load object store, wandb doesn't need `log_artifacts` set to true, right?

File: composer/trainer/trainer.py [hunk 3]
```diff
@@ -532,7 +558,9 @@ class Trainer:
         load_object_store (Union[ObjectStore, LoggerDestination], optional): If the ``load_path`` is in an
             object store (i.e. AWS S3 or Google Cloud Storage), an instance of :class:`.ObjectStore` or
             :class:`.LoggerDestination` which will be used to retreive the checkpoint. Otherwise, if the
-            checkpoint is a local filepath, set to ``None``. Ignored if ``load_path`` is ``None``.
+            checkpoint is a local filepath, set to ``None``. Also, it can be ``None`` is the ``load_path`` is
```
Comment (dakinggg) [inline composer/trainer/trainer.py:561]: "is the" -> "if the"

**Review by dakinggg — COMMENTED** [review]
Lgtm, will approve once docs/examples are updated.

File: composer/trainer/trainer.py (hunk 1 above)
Comment (eracah) [reply composer/trainer/trainer.py]: fixed

File: composer/trainer/trainer.py (hunk 2 above)
Comment (eracah) [reply composer/trainer/trainer.py]: fixed

File: composer/trainer/trainer.py (hunk 3 above)
Comment (eracah) [reply composer/trainer/trainer.py]: fixed

**Review by dakinggg — APPROVED** [review]
LGTM, would still be good to update the main checkpointing docs page with the new options.

**Pull request merged**